### PR TITLE
Adding XOF package.

### PR DESCRIPTION
**Pull request by armfazh**

### CODE REVIEW

File: xof/xof.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	io.Writer
+
+	// Read reads more output from the hash. It returns io.EOF if the limit
+	// has been reached.
+	io.Reader
+
+	// Clone returns a copy of the XOF in its current state.
+	Clone() XOF
+
+	// Reset resets the XOF to its initial state.
```
Comment (wbl) [inline xof/xof.go:26]: While this is copied from the blake2b definition, I had some trouble understanding how this works for keyed ones. Maybe say "discards all data appended by Write" to clarify.

File: xof/xof.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+var xofRegistry = make([]func() XOF, maxID)
+
+func init() {
+	RegisterXOF(SHAKE128, newShake128)
+	RegisterXOF(SHAKE256, newShake256)
+	RegisterXOF(BLAKE2XB, newBlake2xb)
+	RegisterXOF(BLAKE2XS, newBlake2xs)
+}
+
+const maxID = 4
```
Comment (wbl) [inline xof/xof.go:60]: I would add this to the end of the constant definitions.

File: xof/xof.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+func RegisterXOF(x ID, f func() XOF) {
+	if x >= maxID {
+		panic("crypto: RegisterXOF of unknown XOF function")
+	}
+	xofRegistry[x] = f
+}
+
+var xofRegistry = make([]func() XOF, maxID)
+
+func init() {
```
Comment (wbl) [inline xof/xof.go:53]: This makes this package which just defines an interface import these other packages, which makes the whole registration business a bit less useful. Not that the registration was very useful in the first place imho.

**Review by wbl — CHANGES_REQUESTED** [review]

File: xof/xof.go (hunk 3 above)
Comment (armfazh) [reply xof/xof.go:53]: agree, registry not needed at all, changed now.

File: xof/xof.go (hunk 1 above)
Comment (armfazh) [reply xof/xof.go:26]: updated now

**Review by claucece — APPROVED** [review]

**Review by wbl — APPROVED** [review]

**Pull request merged**